### PR TITLE
Add new GLUE example with no Trainer.

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a new GLUE example that does not use the `Trainer`, leveraging [accelerate](https://github.com/huggingface/accelerate) for the distributed training. The necessary instructions are added in the text-classification README.

cc @JetRunner as it should be of interest to you.

### CODE REVIEW

File: examples/text-classification/run_glue_no_trainer.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        help="If passed, pad all samples to `max_length`. Otherwise, dynamic padding is used.",
+    )
+    parser.add_argument(
+        "--model_name_or_path",
+        type=str,
+        help="Path to pretrained model or model identifier from huggingface.co/models.",
+        required=True,
+    )
+    parser.add_argument(
+        "--use_slow_tokenizer",
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:94]: `--use_native/traditional/python/legacy_tokenizer`?
`slow tokenizer` sounds terrible but it's actually not that slow.

BTW, some users told me fast tokenizer sometimes output different results from the native ones. But I can't really reproduce that.

File: examples/text-classification/run_glue_no_trainer.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+            label_list = datasets["train"].unique("label")
+            label_list.sort()  # Let's sort it for determinism
+            num_labels = len(label_list)
+
+    # Load pretrained model and tokenizer
+    #
+    # In distributed training, the .from_pretrained methods guarantee that only one local process can concurrently
+    # download model & vocab.
+    config = AutoConfig.from_pretrained(args.model_name_or_path, num_labels=num_labels, finetuning_task=args.task_name)
+    tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path, use_fast=not args.use_slow_tokenizer)
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:243]: I personally prefer `use_slow` by default.

File: examples/text-classification/run_glue_no_trainer.py [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+            loss = outputs.loss
+            loss = loss / args.gradient_accumulation_steps
+            accelerator.backward(loss)
+            if step % args.gradient_accumulation_steps == 0 or step == len(train_dataloader) - 1:
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
+                progress_bar.update(1)
+                completed_steps += 1
+
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:391]: Logging/saving during training is a kind of necessity for researchers. Should have a `--logging_steps` here and `--saving_steps`.

**Review by JetRunner — COMMENTED** [review]
@sgugger Very impressive! `accelerate` seems very user-friendly and I definitely like it.

I think the arguments should be consistent with `transformers==2.11.0` (which is pretty close to Google's old `run_glue.py`). Not because I'm nostalgic but I think we should respect researchers' customs and try not to go against them. Also, I think all features in the old script are necessary and we shouldn't really cut them out. Just have something compatible with the latest version of `transformers` but has the same parameters as the old `run_glue.py` is the best IMO. **Researchers typically don't care about fancy features like picking the best checkpoint automatically but let's at least don't have fewer features than Google's run_glue.**

 Also, [making predictions](https://github.com/google-research/bert/blob/master/run_classifier.py#L944) (`--do_predict`) is critical for researchers so I kinda cannot understand why it's supported in Trainer-version `run_glue` but not here.

**Comment by JetRunner** [comment]
This kind of logging is very useful for researchers. Let's add them back?

https://github.com/google-research/bert/blob/master/run_classifier.py#L871

**Comment by JetRunner** [comment]
In a nutshell, I'll burst into tears if we can just have Google's `run_classifier.py` back but with `accelerate` :)

File: examples/text-classification/run_glue_no_trainer.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+    parser.add_argument(
+        "--lr_scheduler_type",
+        type=SchedulerType,
+        default="linear",
+        help="The scheduler type to use.",
+        choices=["linear", "cosine", "cosine_with_restarts", "polynomial", "constant", "constant_with_warmup"],
+    )
+    parser.add_argument(
+        "--no_correct_bias_in_adam", action="store_true", help="If passed, no bias correction is applied in AdamW."
+    )
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:139]: Not really necessary to me? Researchers can just tweak it manually if they need that.

File: examples/text-classification/run_glue_no_trainer.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    label_to_id = None
+    if (
+        model.config.label2id != PretrainedConfig(num_labels=num_labels).label2id
+        and args.task_name is not None
+        and is_regression
+    ):
+        # Some have all caps in their config, some don't.
+        label_name_to_id = {k.lower(): v for k, v in model.config.label2id.items()}
+        if list(sorted(label_name_to_id.keys())) == list(sorted(label_list)):
+            label_to_id = {i: label_name_to_id[label_list[i]] for i in range(num_labels)}
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:274]: Let's add a warning too? Users should know their label assignment is not the default one.

File: examples/text-classification/run_glue_no_trainer.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+                sentence1_key, sentence2_key = non_label_column_names[:2]
+            else:
+                sentence1_key, sentence2_key = non_label_column_names[0], None
+
+    # Some models have set the order of the labels to use, so let's make sure we do use it.
+    label_to_id = None
+    if (
+        model.config.label2id != PretrainedConfig(num_labels=num_labels).label2id
+        and args.task_name is not None
+        and is_regression
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:269]: `and is_regression`? Do you mean `and not is_regression`?

File: examples/text-classification/run_glue_no_trainer.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+    ):
+        # Some have all caps in their config, some don't.
+        label_name_to_id = {k.lower(): v for k, v in model.config.label2id.items()}
+        if list(sorted(label_name_to_id.keys())) == list(sorted(label_list)):
+            label_to_id = {i: label_name_to_id[label_list[i]] for i in range(num_labels)}
+        else:
+            logger.warn(
+                "Your model seems to have been trained with labels, but they don't match the dataset: ",
+                f"model labels: {list(sorted(label_name_to_id.keys()))}, dataset labels: {list(sorted(label_list))}."
+                "\nIgnoring the model labels as a result.",
```
Comment (JetRunner) [inline examples/text-classification/run_glue_no_trainer.py:279]: Add `Please make sure the label assignment is consistent with {list(sorted(label_list))}.`? Then we should let the users know how to overwrite the label assignment?

**Review by JetRunner — COMMENTED** [review]
Some additional comments.

**Comment by JetRunner** [comment]
Maybe we should tag other researchers to give some feedback. cc @VictorSanh @TevenLeScao

File: examples/text-classification/run_glue_no_trainer.py (hunk 2 above)
Comment (sgugger) [reply examples/text-classification/run_glue_no_trainer.py:240]: Fast is the default of the library since 4.0.0.

File: examples/text-classification/run_glue_no_trainer.py (hunk 1 above)
Comment (sgugger) [reply examples/text-classification/run_glue_no_trainer.py:94]: I believe the different results are actually bug fixes but it's hard to say without any example.

**Comment by sgugger** [comment]
Addressed most of your comments except the logging/saving steps. I do not have time to add this right now, so I suggest we merge the current version and someone from the community can finish it.

**Review by JetRunner — APPROVED** [review]
LGTM! I can make some improvements incl. logging/saving steps.

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @sgugger. Looking forward to future improvements by @JetRunner too!
